utils/s3/client.cc

```diff
@@ -1092,6 +1093,118 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece, as));
 }
 
+class client::chunked_download_source final : public seastar::data_source_impl {
+    struct claimed_buffer {
+        temporary_buffer<char> _buffer;
+        semaphore_units<> _claimed_memory;
+        s3_clock::duration _consumption_time;
+        claimed_buffer(temporary_buffer<char>&& buf, semaphore_units<>&& claimed_memory, s3_clock::duration consumption_time)
+            : _buffer(std::move(buf)), _claimed_memory(std::move(claimed_memory)), _consumption_time(consumption_time) {}
+    };
+    shared_ptr<client> _client;
+    sstring _object_name;
+    seastar::abort_source* _as;
+    range _range;
+    static constexpr size_t _max_buffers_size = 5_MiB;
+    std::deque<claimed_buffer> _buffers;
+    size_t _buffers_size = 0;
+    bool _is_finished = false;
+    seastar::condition_variable _buffer_cv;
+    future<> _filling_fiber = make_ready_future<>();
+
+    future<> make_filling_fiber() {
+        while (!_is_finished) {
+            try {
+                s3l.trace("Fiber starts cycle for object '{}'", _object_name);
+
+                if (_buffers_size >= _max_buffers_size) {
+                    co_await _buffer_cv.when([this] { return _buffers_size < _max_buffers_size; });
+                }
+
+                if (_is_finished) {
+                    s3l.trace("Abandoning fiber for object '{}'", _object_name);
+                    co_return;
+                }
+
+                auto req = http::request::make("GET", _client->_host, _object_name);
+                req._headers["Range"] = format_range_header(_range);
+                s3l.trace("Fiber for object '{}' will make HTTP request within range {}-{}", _object_name, _range.off, _range.len);
+                co_await _client->make_request(
+                    std::move(req),
+                    [this](const http::reply& reply, input_stream<char>&& in_) mutable -> future<> {
+                        if (reply._status != http::reply::status_type::ok && reply._status != http::reply::status_type::partial_content) {
+                            s3l.warn("Fiber for object '{}' failed: {}. Exiting", _object_name, reply._status);
+                            throw httpd::unexpected_status_error(reply._status);
+                        }
+                        auto in = std::move(in_);
+                        while (_buffers_size < _max_buffers_size && !_is_finished) {
```

Comment:
Probably need a debug-level message that the fiber is abandoning reply body reading early

Reply:
Ah, http client has such trace-level message already